Removed observer classes

AtlantiaKing opened the pull request:
Removed the Observer class and the templated TObserver class
Added AddListener and RemoveListener methods to the Subject and TSubject class with member functions as parameters

AtlantiaKing committed:
Removed Observer and added member functions to Subject

Removed Observer and TObserver class
Added AddListener and RemoveListener functions for member functions and const member functions to the Subject class
Replaced Observer logic with member function delegates in the whole engine

AtlantiaKing committed:
Merge branch 'main' into remove-observer

AtlantiaKing committed:
Removed observer in MeshRenderer class

AtlantiaKing committed:
Uncommented LeapClient

AtlantiaKing committed:
Fixed MeshRenderer never calling AddListener

LeeVangraefschepe commented on Utils/Subject.h:
```diff
@@ -1,104 +1,270 @@
+		struct ImplMemberFunctionWrapper final : public MemberFunctionWrapper
+		{
+			ImplMemberFunctionWrapper(T* _pThis, void(T::* _pFunction)())
+				: pThis{ _pThis }, pFunction{ _pFunction }, id{ ReflectionUtils::GenerateTypenameHash<T>() }
```
Initialization order will not be executed like this (same for Const class type).
id,
pThis,
pFunctuin

LeeVangraefschepe commented on Utils/Subject.h:
```diff
@@ -1,104 +1,270 @@
+			std::erase(m_Functions, pCallback);
 		}
 
 		void Notify()
```
Notify call can be set const now

LeeVangraefschepe commented on Utils/Subject.h:
```diff
@@ -1,104 +1,270 @@
+			virtual ~ImplMemberFunctionWrapper() = default;
 
-		void AddListener(Observer* observer)
+			unsigned int id{};
```
All member variables publicly exposed.
I know its internal but you also made a GetId() function for a publicly exposed id member var.

LeeVangraefschepe commented on Utils/Subject.h:
```diff
@@ -1,104 +1,270 @@
+		void RemoveListener(TClass* pThis, void(TClass::* pFunction)(const T&))
+		{
+			constexpr unsigned int classId{ ReflectionUtils::GenerateTypenameHash<TClass>() };
+			for (int i{ static_cast<int>(m_pMemberFunctions.size()) - 1 }; i >= 0; --i)
```
Cache vector size. Every for tick the size will get recalculated.

AtlantiaKing replied:
Vector size is only used for initializing the for index

LeeVangraefschepe commented on Networking/Leap/LeapClient.cpp:
```diff
@@ -50,7 +50,7 @@ leap::networking::LeapClient::LeapClient(int port, const std::string& serverIp,
         std::stringstream ss{};
         ss << "Error connecting to server: " << WSAGetLastError() << std::endl;
         Debug::LogError(ss.str());
-        
```
Clear out empty change (merge conflict will happen otherwise)

LeeVangraefschepe reviewed (changes requested):
Good cleanup you removed all #include "Observer.h" from every header file it was used in :)
Also another amazing usage for the new reflection utils.
Check comments for the review.

AtlantiaKing committed:
Fixed tabs being removed in the LeapClient class

This commit has been made on a branch while LeapClient.cpp was being edited on another branch, this chance was needed to prevent merge conflicts

AtlantiaKing committed:
Code review Subject class

Moved id variable from classes that derive from the MemberFunctionWrapper struct to the base struct MemberFunctionWrapper
Removed GetId() method from the MemberFunctionWrapper struct and its derived structs
Added constructo to the MemberFunctionWrapper struct that takes an unsigned int as parameter
The notify method in the Subject and TSubject class are now const

LeeVangraefschepe reviewed (approved):
Code feedback is applied.
